synapse/push/pusherpool.py

```diff
@@ -128,6 +127,22 @@ async def add_or_update_pusher(
         # stream ordering, so it will process pushes from this point onwards.
         last_stream_ordering = self.store.get_room_max_stream_ordering()
 
+        # Before we actually persist the pusher, we check if the user already has one
+        # this app ID and pushkey. If so, we want to keep the access token and device ID
+        # in place, since this could be one device modifying (e.g. enabling/disabling)
+        # another device's pusher.
+        # XXX(quenting): Even though we're not persisting the access_token_id for new
+        # pushers anymore, we still need to copy existing access_token_ids over when
+        # updating a pusher, in case the "set_device_id_for_pushers" background update
+        # hasn't run yet.
```

Comment:
So this was introduced by [MSC3881](https://github.com/matrix-org/matrix-spec-proposals/pull/3881). With that MSC, you might be updating a pusher that isn't yours, and in that case you want to keep the access_token_id/device_id, and not replace it with the one from the current token.

The only reason we're also carrying the access_token_id here, is that this might happen while the background update hasn't happened yet, and we don't want to break those

Reply:
> With that MSC, you might be updating a pusher that isn't yours, and in that case you want to keep the access_token_id/device_id, and not replace it with the one from the current token.

So the `device_id` provided to this function is *always* the device performing the action?

Reply:
It's a bit counterintuitive, but yes. That behaviour was introduced in #13799